### PR TITLE
[20269][Structured Streaming] add class 'JavaWordCountProducer' to 'provide java word count producer'.

### PR DESCRIPTION
**Pull request by guoxiaolongzte**
## What changes were proposed in this pull request?

1.run example of streaming kafka,currently missing java word count producer,not conducive to java developers to learn and test.
add a class JavaKafkaWordCountProducer.

2.run example  of JavaKafkaWordCount.I find no java word count producer.
run example of KafkaWordCount.I find have scala word count producer.
I think we should provide the corresponding example code to facilitate java developers to learn and test.

3.My project team develops spark applications,basically with java statements and java API.

4.test case
  1)java kafka word count producer,generate data for run example of JavaKafkaWordCount.the class of JavaKafkaWordCountProducer  add by me.
  spark-submit ... --class org.apache.spark.examples.streaming.JavaKafkaWordCountProducer  --jars  /home/gxl/spark/libext/kafka-clients-0.8.2.1.jar /home/gxl/spark/examples/jars/spark-examples_2.11-2.1.0.jar localhost:9092 topic1 3 4

  2)java kafka streaming example
spark-submit ... --class org.apache.spark.examples.streaming.JavaKafkaWordCount  --jars /home/gxl/kafka/kafka_2.11-0.10.2.0/libs/kafka-clients-0.10.2.0.jar --jars /home/gxl/spark/libext/kafka_2.11-0.8.2.1.jar /home/gxl/spark/examples/jars/spark-examples_2.11-2.1.0.jar localhost:2181 topic1 topic1 2

## How was this patch tested?

manual tests

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by srowen** [comment]
The code style needs to be fixed, and the title. What example is this based on ? this is the kind of info that you should put in a pull request.

**Comment by guoxiaolongzte** [comment]
Title,PR description and motive, has been modified.

File: examples/src/main/java/org/apache/spark/examples/streaming/JavaKafkaWordCountProducer.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+ * Produces some random words between 1 and 10
+ * Usage: JavaKafkaWordCountProducer <metadataBrokerList> <topic> <messagesPerSec> <wordsPerMessage>
+ *   <metadataBrokerList> is a list of one or more kafka broker servers
+ *   <topic> is a kafka topic to consume from
+ *   <messagesPerSec> is the number of messages generated per second
+ *   <wordsPerMessage> is the number of words per message
+ */
+public class JavaKafkaWordCountProducer {
+    public static void main(String[] args) {
+        if (args.length < 4) {
```
Comment (HyukjinKwon) [inline examples/src/main/java/org/apache/spark/examples/streaming/JavaKafkaWordCountProducer.java]: We follow 2 space indentation for Java. Please review "Code Style Guide" in http://spark.apache.org/contributing.html

**Comment by guoxiaolongzte** [comment]
Sorry,spark java code style is different from the style of my project team.Now I know, and have been fixed.


Use 2-space indentation in general. For function declarations, use 4 space indentation for its parameters when they don’t fit in a single line.

**Comment by jerryshao** [comment]
What is the purpose of adding this example? I think we already have a `KafkaWordCountProducer` for the convenience of Kafka streaming example, and we could use that to send events to Kafka. I think there's no need to add a Java counterpart, since this class is only for the test convenience and has no direct relation to Spark.

**Comment by guoxiaolongzte** [comment]
When a user use spark to develop a stream&kafka  application, he first wants to find and learn  example program in 'spark \ examples \ src \ main \ java \ org \ apache \ spark \ examples \ streaming' directory, will not see Scala directory any example of the program, so the example program has limited help to him. If you add a Java JavaKafkaWordCountProducer, for him to completely solve the problem.

Because since this class is only for the test convenience ,so i think it  has direct relation to Spark.

**Comment by jerryshao** [comment]
I would say this is not a Spark program, it is absolutely a Kafka producer code. To maintain a Kafka Producer example in Spark is not a good choice, this is a legacy code. Because of the API changes of Kafka, the example here will become obsolete and maintenance overhead (I fixed the deprecated APIs here once shifting to 0.8.2). So instead of adding Java counterpart, I would rather to remove it. (Just my two cents).

**Comment by guoxiaolongzte** [comment]
Because of the API changes of Kafka, we do not want to delete it, but to maintain and modify it.
Although it is absolutely a Kafka producer code, but this is part of spark streaming, it is also very important.

**Comment by srowen** [comment]
I think I'd go with @jerryshao and not add it if it's an old, auxiliary example anyway

**Comment by guoxiaolongzte** [comment]
@jerryshao 

import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
The above three classes in kafka-clients-0.10.0.1.jar。The example here will not become obsolete.

**Comment by srowen** [comment]
I think this is not worth bothering with for the reasons above, and can be closed.

**Comment by guoxiaolongzte** [comment]
@jerryshao
Excuse me,I would like to ask, these examples of Java examples, why use some of the Scala to write it?thank you!




    SparkConf sparkConf = new SparkConf().setAppName("**JavaKafkaWordCount**");
    // Create the context with 2 seconds batch size
    JavaStreamingContext jssc = new JavaStreamingContext(sparkConf, new Duration(2000));

    int numThreads = Integer.parseInt(args[3]);
    Map<String, Integer> topicMap = new HashMap<>();
    String[] topics = args[2].split(",");
    for (String topic: topics) {
      topicMap.put(topic, numThreads);
    }

    JavaPairReceiverInputDStream<String, String> messages =
            KafkaUtils.createStream(jssc, args[0], args[1], topicMap);

    JavaDStream<String> lines = messages.map**(Tuple2::_2);**

**Comment by jerryshao** [comment]
What's the meaning of "some of the Scala"?

**Comment by guoxiaolongzte** [comment]
eg
   messages.map(Tuple2::_2) is the Scala grammar code.But JavaKafkaWordCount is a java class.
   I think the Java example program should be fully implemented in Java code.
  What do you think?

**Comment by jerryshao** [comment]
You mean this [line](https://github.com/apache/spark/blob/master/examples/src/main/java/org/apache/spark/examples/streaming/JavaKafkaWordCount.java#L76)?

It's because our KafkInputDStream signature is `ReceiverInputDStream[(K, V)]`, in which the Type is `(K, V)`,  this `(K, V)` is Scala's Tuple2. That's why in the Java code we need to call Scala's API.

**Comment by guoxiaolongzte** [comment]
JavaDStream<String> lines = messages.map(new Function<Tuple2<String, String>, String>() {
      @Override
      public String call(Tuple2<String, String> tuple2) {
        return tuple2._2();
      }
    });

    JavaDStream<String> words = lines.flatMap(new FlatMapFunction<String, String>() {
      @Override
      public Iterator<String> call(String x) {
        return Arrays.asList(SPACE.split(x)).iterator();
      }
    });

    JavaPairDStream<String, Integer> wordCounts = words.mapToPair(
      new PairFunction<String, String, Integer>() {
        @Override
        public Tuple2<String, Integer> call(String s) {
          return new Tuple2<>(s, 1);
        }
      }).reduceByKey(new Function2<Integer, Integer, Integer>() {
        @Override
        public Integer call(Integer i1, Integer i2) {
          return i1 + i2;
        }
      });



**The above code version in 2.1.0 release.It is only recently changed to Scala grammar in apache master. I want to know why.Thank you!**@jerryshao

**Comment by jerryshao** [comment]
It is just Java8 lambda function, nothing related to Scala...

**Comment by srowen** [comment]
In some cases you have to call a Scala API from Java and need to create Scala-specific classes like Tuple2.

**Comment by guoxiaolongzte** [comment]
@srowen
@jerryshao 
I understand, thank you.
I changed the code in my project to keep the program consistent with the latest example.

**Pull request closed without merging**